r/R/dplyr.R

```diff
@@ -216,6 +215,52 @@ column_select <- function(.data, ..., .FUN = vars_select) {
   .data
 }
 
+relocate.arrow_dplyr_query <- function(.data, ..., .before = NULL, .after = NULL) {
```

Comment:
The code here is adapted from `dplyr::relocate.data.frame()`. https://github.com/tidyverse/dplyr/blob/614389a391504ee17708aca2cce3d92b123ac449/R/relocate.R#L53-L90
Should I add a note here or elsewhere to cite this?

Reply:
Yes, here is fine I think, assuming that you're adapting and not just copying. If it's a significant enough copy, then we may need to add a note to the LICENSE file (IDK what the threshold is).

Side note: I believe if `[<-` were defined for `arrow_dplyr_query`, then `relocate.arrow_dplyr_query <- dplyr:::relocate.data.frame` may just work (in which case we would want to make a PR to make that `relocate.default`). Then we wouldn't need to copy the code. It might be worth implementing that method, at least enough for column permutation (which is just operating on `$selected_columns`).

Reply:
The changes were more than trivial. Here's what changed: https://gist.github.com/ianmcook/0eba58ce40077fff6d02f73b459f903d/revisions

Reply:
Ok. I did also test out calling `dplyr:::relocate.data.frame(.data$selected_columns, ...)` and that works except for the one `ncol()` reference (interestingly, patching that to be `length()` does mean you can just reuse the function, and `length()` also happens to work on `data.frame`, so maybe that's the angle to take).

Reply:
Hmm, interesting, or we could assign a class to `.data$selected_columns` and implement the `dim` generic for it

Reply:
After thinking about this more, direct reuse of  `dplyr:::relocate.data.frame` (without copying) is an intriguing possibility but it would introduce a dependency on internal implementation details of dplyr that I think we ought to avoid. The MIT license (under which dplyr is licensed) permits this type of reuse (with attribution, which I will add in a comment).

Reply:
Ok, add a comment here referencing the original version, and perhaps also link to https://github.com/tidyverse/dplyr/issues/5829 in the comment

Reply:
Added in 35cf9c553b47726594089cf97cbd5af4f1049da6